### ISSUE TITLE
ed25519 subdirecotry and code/dependencies

### ISSUE THREAD

**Issue by dmitris** [yahoo/coname#61]
Currently there is a ed25519 subdirectory, but there are also imports of github.com/agl/ed25519 (ex. https://github.com/agl/ed25519/blob/master/ed25519.go#L17 , https://github.com/yahoo/coname/blob/master/keyserver/server.go#L36, https://github.com/yahoo/coname/blob/master/keyserver/deploy/make_config.go#L38).  The reason for both "vendored" (embedded) directory and imported does not seem obvious (possibly historic or related to the codebase evolution?).  Since Adam put his ed25519 code into golang.org/x/crypto/ed25519 - see https://github.com/golang/crypto/commit/c9aef117f53e16ed3a6cc3d93cd357dbf2005065 - should we standardize all references and imports to that "official" version?  If there are no objections in general, I can send a PR for the differences in API etc.

@maditya

**Comment by adon-at-work** [comment]
cc-ing @andres-erbsen and @daniel-ziegler too. thanks :)

**Comment by andres-erbsen** [comment]
The vendored ed25519 subdirectory implements functionality that is not present in AGL's code. The added functionality is only required for implementing the VRF. I think it would make sense to have only the VRF implementation import our copy of the package, and everything else would import the go.crypto version to receive automatic updates.

**Comment by bford** [comment]
As part of our work we have generalized AGL's port of the ed25519 code to support a fairly complete algebraic interface for both points and scalars, supporting all the operations needed not only for VRFs but also zero-knowledge proofs, Neff shuffles, etc.  Would it perhaps make sense to have the VRF code use that instead of "yet another" specially-customized copy of the ed25519 arithmetic code?

- Point and scalar interfaces (the latter is currently called "Secret" but we're planning to rename to "Scalar"): https://godoc.org/github.com/dedis/crypto/abstract

- Implementation for ed25519 based on AGL's port of DJB's code: https://godoc.org/github.com/dedis/crypto/edwards/ed25519

If you think this makes sense but there are things about our crypto library or the arithmetic interfaces you don't like, please let us know - we're working on some cleanups and trimming in advance of a "v1.0 stable" release we're planning, but nothing is frozen yet.

**Comment by andres-erbsen** [comment]
I agree that the having yet another copy of ed25519 arithmetic code is
unfortunate, but I am currently judging the defect-risk due to duplication of
code to be smaller than the risk of potentially unsound abstraction. I am all
for proper use of abstraction in cryptographic system design, but when it comes
to implementation, focusing on specific primitives has a much better correctness
track record.

Three things that stuck out to me when I looked into `dedis/crypto` for 15
minutes:

- <https://github.com/dedis/crypto/blob/master/edwards/ed25519/point.go#L238>
  returns that Curve25519 is a prime-order group, which is incorrect. The base
point has large prime order, and thus can be used in many situations where NIST
curves can, but not all of them. It is not the same abstraction.
- <https://github.com/dedis/crypto/blob/3a88a8f7308158f184a08c853b4ac9c17a5bce35/nist/int.go#L183>
  uses `big.Int` which is not constant time. These methods are plausibly used
with ECDSA secrets and nonces for which side-channel security is absolutely
critical. Generic implementations with runtime independent of secret inputs are
possible, but have not turned out that well so far (e.g.
<https://eprint.iacr.org/2016/594.pdf>). The Go crypto library seems to be
favoring specialized constant-time implementations.
- <https://github.com/dedis/crypto/blob/3a88a8f7308158f184a08c853b4ac9c17a5bce35/edwards/ed25519/point.go#L206>
  implements Curve25519 point multiplication in variable time. Similar caveats
apply.

I am not convinced that pursuing a fully general and correct implementation of
the abstraction proposed by `dedis/crypto/abstract` is a good idea. Doing the
design at that abstraction level makes a lot of sense, but when implementing an
actual system, the details matter a lot -- to the extent that reasoning about
specific curves seems to make a lot of sense. Different groups are not
interchangeable in most settings. Just to name a few points of consideration:
some are prime-order and some are not, some support cofactor-removing encoding
and some do not, some have reasonably performant complete addition formulas and
some do not, the methods for hashing into different curves vary in terms of what
properties they provide, etc. These properties bubble up to the protocol level:
I can't think of what one would need to do to a NIST curve to build a drop-in
replacement for X25519 key agreement.

A larger set of abstractions (e.g. `Group` and `PrimeOrderGroup`,
`UnmarshalFrom` and `MaybeUnmarshalFrom` and `CanonicalUnmarshalFrom`) can
probably be defined in a sound and consistent manner to ensure that
PrimeOrderGroup points are always in the correct subgroup of the elliptic curve
and so on. This would involve specifying things like whether "prime-order group
points" include the identity (which is in the prime-order subgroup but does not
have prime order); currently
<https://github.com/dedis/crypto/blob/322235ade1054c3eb2590f6a6a6c719ae6b011ff/edwards/curve.go#L274>
and
<https://github.com/yahoo/coname/blob/master/ed25519/edwards25519/edwards25519.go#L1095>
behave differently. I think a rigorous analysis and documentation of both the
properties of different (curve, point format, decoding procedure) combinations
would be valuable in the long run (and even more so after the requirements of
protocols start being documented as well). After reading `dedis/crypto`
a bit more, it seems to me that the project is already going in this direction
(perhaps just the `ed25519` subdirectory is behind?). I would also advocate for
duplicating methods based on whether they are allowed to leak their input
through timing: `Mul` would be constant-time, `MulByPublicScalar` could make use
of sliding-window optimizations.

In principle, a merger of `coname/ed25519` and `dedis/crypto/edwards` is
desirable. However, based on the experience that concrete implementations tend
to just avoid a significant class of mistakes, I think we should make very sure
these issues are addressed in a systematic manner.

**Comment by bford** [comment]
Hi Andres, thanks for your thoughts and review of the code.  I agree that building a good [semi-]abstract API for curve arithmetic is not trivial, but I think that means we need to put in the effort go get it right, not just give up on the task.  Creating - and trying to maintain over the long-term - separate and slightly-hacked implementations of all the basic X25519 curve arithmetic for many different crypto schemes (Diffie-Hellman key exchange, signatures, VRFs, zero-knowledge proofs, Neff shuffles, etc.) has plenty of real-world costs as well, in terms of software engineering and security risks, e.g., as bugs found and fixed in one copy of the code fail to propagate to others, etc.

Regarding the specific issues you brought up:

* It looks like you misunderstood the "sense" of the return value of the PrimeOrder method: its return does not and is not supposed to indicate that the *curve* is prime-order, but rather whether or not the group or subgroup *in use* is prime-order, as defined by the base-point returned by Base().  So it should probably have been more descriptively named "BaseIsPrimeOrder" or something like that for clarity.  I believe the one and only place it is currently used is in the generic curve arithmetic testing code, to decide whether operations like modular inverse and square-root can be expected to work or not and hence whether those operations should be tested at all.  It's understandable that you missed that subtlety of this particular method, since it's rarely-used, not documented in the ed25519 implementation, and may not be sufficiently clearly documented anywhere; that's some of the type of issues we're trying to find and fix during the currently in-progress cleanup work.

* Regarding constant-time versus variable-time implementations, I agree it is extremely important to take care of this and make sure the difference is clear in the API.  The ed25519 scalar code's use of nist.Int was a temporary hack, and fixing it has been on my TODO list for a while: you may have noticed that the ScMulAdd code is still in the package although currently unused, and a long-outstanding TODO item has been to decompose that to separate Mul from Add and use it to provide a proper constant-time implementation of the scalar arithmetic.

* Similarly, regarding the current point.Mul() implementation, when you say that is currently variable-time I assume you're just referring to the fact that it uses nist.Int.V.Bytes() to extract the byte-representation of the secret, whose result-length and hence timing can vary depending on the contents of the integer?  That's just an again temporary artifact of the same hack above; scalars aren't supposed to and shouldn't use nist.Int or big.Int, and we just need to (finally) fix that.  It was always in the plans to do that before even contemplating any production use of the code.  Everything else in point.Mul should already be constant-time and is derived directly from AGL's port of DJB's code, unless I'm missing or forgot something else - let me know if you say anything else I might have missed.

But another general API issue of course is that it's sometimes useful to have possibly faster vartime implementations available for operations that support it, like scalar multiplication, and another long-standing TODO item has been to figure out the right API way to expose that.  We could add ...Vartime() versions of some or all operations on the interface, but that seems clunky.  Perhaps a better alternative may be just to add a SetVartime(bool vartime) method that sets an internal state flag indicating whether faster vartime implementations of the operations are allowed to be used if they're available.  Obviously this should always default to false for safety: i.e., no vartime operations should be used unless the caller has indicated having thought about it and explicitly requested it.

And certainly the way the API exposes the difference between prime-order and non-prime-order curves and points needs to be improved in the current API; specific suggestions are very welcome.  It wasn't really my intention to try to "hide" the difference between prime-order and cofactored curves, since indeed the difference is often important and must be considered in design; it just wasn't clear what the best way to expose the difference is.  So that's another ongoing TODO item we need to work on.

Cheers
Bryan

**Comment by andres-erbsen** [comment]
I sincerely hope that `dedis/crypto` breaks the pattern of concrete
implementations outlasting modular ones in terms of correctness.
I continue to be intimidated by the scope and delicacy of the targeted
abstractions, but I wish you best of luck; I would happily review a design document for
adding a VRF like this into `dedis/crypto`.
